[WIP] Blood Overhaul + Synthetics

Poojawa opened the pull request:
:cl: Poojawa
add: Synthetics blood is added, it is a artificial blood replacement compatible with all species. Made in the limb grower with regular blood of any type.
tweak: Species have been given their own brand of blood, colored differently. This enables IPCs and Slimes to suffer from Virology just like everyone else. Jelly blood can be mixed with toxin for a 2:1 slimejelly production.
add: Everyone bleeds now, no one is NOBLOOD anymore.
tweak: Blood is now DNA and RGB'd, adapting to the mix within it.
fix: fixes spawn paths for roundstart slimes and IPCs, admins can spawn these now.
refractor: Gibspawners have better code now.
add: Mobs have their own unique type of gib spawn now. based on species. IPCs explode into robot gibs, Xenos into xenomorph gibs. They also have unique gib types that provide a DNA sample.

/:cl:

Current Issues: 
~~footprints are invisible~~, they do not take colors. 
Footprints of shoe-less mobs do not take any blood material. Needed for non-shoe/dragging prints
~~Gib overlays are not setting themselves to null for color, so they're full bright of whatever the blood consists of.~~
Blood types need to be listed for color mixing, but Idk how to do that without adding a list to every atom. ree. Currently hijacking forensics components because that's where I'd be getting all the DNA data from anyway. Color mixing and handling already done. we're just waiting for plug-and-play

TODO: 

- [ ] Add functionality to limb grower to make synth limbs from synth blood
- [x] ensure RGB gibs properly layer and absent themselves from color
- [ ] probably make a fancy Tgui UI for the limb grower
- [ ] adapt footsteps to take into account body types, mammals with paws if digitigrade, etc.

Poojawa committed:
Synthetics blood and greyscale blood effects

Poojawa committed:
compile fixes except we're memes still

why yes, I'm making the bloods all subtypes of each other at this point. I should probably change the way stuff is colored. but, meh. effort.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
pretty sure I'm onto it

Poojawa committed:
greyscale digi shoe blood

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Synthetic and RNG blood

Poojawa committed:
oh god oh fuck oh god it's snowballing

Poojawa committed:
Some slight fixes

Poojawa committed:
system's mostly working

need to polish out the get_blood_dna_list runtime

Poojawa committed:
what to do when footprints are being a bitch.

why, we get rid of them :honk:

Poojawa committed:
I need to do something else for a bit pending reviews

Poojawa committed:
whatever Travis, it isn't even enabled, fuck off

Poojawa committed:
makes travis shutup

Poojawa committed:
I really dunno at this point, weh

Poojawa committed:
go away Travis

Trilbyspaceclone commented on code/modules/reagents/chemistry/recipes/medicine.dm:
```diff
@@ -233,6 +233,12 @@
 	results = list("regen_jelly" = 2)
 	required_reagents = list("tricordrazine" = 1, "slimejelly" = 1)
 
+/datum/chemical_reaction/jelly_convert
```
This seems overly deadly and rather op to heal slimes and to harm non-slimes...

Poojawa replied:
????

No? reagents don't react with blood. this is purely so slimejelly can be produced.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
adjust and reset things

Poojawa committed:
well we have footprints now, xenoblood back to blood subtype.

Poojawa committed:
progress, just need to fiddle with footprints again

Poojawa committed:
get blood id is my foe now.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
okay we're starting to pass our blood type to bloodmeme now

STILL JUST FUCKING FOOTSTEPS LEFT THO GOD FUCKING DAMNIT

Poojawa committed:
Fixes Xenomorph blood related travis errors

Poojawa committed:
Ok this time all of them are nitpicked

Poojawa committed:
duplicate definitions? really? wtf

Poojawa committed:
it's the one part that I keep saying I just need to get around to

Poojawa committed:
Update human_movement.dm

Poojawa committed:
u h g. I hate this

Poojawa committed:
I think I got it.

Also gibs should be better too.

Poojawa committed:
as it turns out, Creatures don't have minds

Poojawa committed:
I don't even fucking know anymore

Poojawa committed:
just need the stupid component to work properly now

Poojawa committed:
port tg 42691 for component transfer enforcement

Poojawa committed:
map editing is nicer with notepad ++ at least

Poojawa committed:
do we even use this shitty away mission

ninjanomnom commented on code/game/objects/effects/decals/cleanable/humans.dm:
```diff
@@ -1,194 +1,216 @@
+/obj/effect/decal/cleanable/blood/update_icon()
+	GET_COMPONENT(F, /datum/component/forensics)
+	if(F && istype(F))
+		blood_color = F.blood_mix_color
```
In case you don't know me I'm basically the primary datum component system maintainer on tg, greetings. I've been summoned here to try and save some components/signals from abuse.

You should not be using GetComponent except in rare cases where an object is basically owned by said component. Even then it is questionable. Take a look at components like squeak for a good example of how to isolate functionality.

You can replace the above code with code like the following:
```
/update_color()
  if(SEND_SIGNAL(COMSIG_UPDATE_COLOR) & COMPONENT_BLOCK_UPDATE_COLOR)
    return

[Inside component receiver of COMSIG_UPDATE_COLOR]
  parent.color = blood_mix_color
  return COMPONENT_BLOCK_UPDATE_COLOR
```

Components should be the holders of all logic related to their own feature. External code to forensics/whatever should not be in charge of how that component behaves. Do not use components as data holders.

If you ever need help with components or signals in any way feel free to pester me, using them requires a different way of going about problems and it can be difficult to pick up until it clicks.

ninjanomnom replied:
The benefits here by the way are many. You make it easier to maintain by making all code related to a functionality by containing the code in the object controlling that functionality. You make it easier to extend or modify that functionality for the same reason. You avoid bugs by not caring what the component is applied to or what its parent is doing. You even get some significant performance benefits by not using the fairly costly GetComponent call or checking if the component exists after.

As a side note you never need to istype the return value.

Poojawa replied:
I'll blame Bhjin for suggesting the istype bit.

As I said, I very much appreciate this write up, even if I'm still a bit too dense to really understand it. most of the component bad-touch was my growing increasingly desperate in a system that just doesn't have enough comparable implementations to what I'd want at an easy at-a-glance sort of learning.

ninjanomnom replied:
I'd suggest you take a look at the more self contained components like, thermite, squeak, and waddle for examples

Poojawa committed:
,

Poojawa committed:
Update wildwest.dmm

Poojawa committed:
Blood colors happens to footprints now.

Different prints not working. splatter effect not working, I'm just sick of looking at this right now.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
reverts forensics components. If I can't figure it out, MURDER IT

Poojawa committed:
oh right.

Poojawa committed:
aaaaAAAAAAAA

NOW WHY DO I HAVE INFINITE FOOTPRINTS AAAAAAAAAAAAA

Poojawa committed:
eh, was going somewhere with this

IPC gibspawner, looking into infinite footprints

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
lel

Poojawa committed:
fuckit, I'm just gunna wholesale port from baycode I guess

Poojawa committed:
these too please thanks

Poojawa committed:
Revert "these too please thanks"

This reverts commit bc80c036791d16bc33d3d841d212003b31c7e102.

Poojawa committed:
Revert "fuckit, I'm just gunna wholesale port from baycode I guess"

This reverts commit 11d10d38052d7cd2077a5b698f4c555b01780354.

Poojawa committed:
Revert "lel"

This reverts commit 9e01a078ec5f8d96e1cb7d9b2c882f1e46ce03a1.

Poojawa committed:
KISS principal

Poojawa committed:
just need to fine tweak gibs and we're gucci

Poojawa committed:
Gibs spawn correctly. Just need to figure out the overlays now

Poojawa committed:
god I can't believe I'm matrixing gib overlays. save me PLEASE

Poojawa committed:
NYEH. fuckin' memes
